File: 322.coin-change.py
#
# @lc app=leetcode id=322 lang=python3
#
# [322] Coin Change
#

# @lc code=start
from typing import *
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [0]
        dp.extend([-1]* (amount))
        for i in range(1,amount + 1 ):
            min_value = -1
            for coin in coins:
                if not i -coin < 0 and not dp[i-coin] == -1 :
                    if min_value == -1:
                        min_value = dp[i-coin] + 1 
                        continue
                    if  dp[i-coin] + 1  <  min_value:
                        min_value = dp[i-coin] + 1 
            dp[i]=min_value
        print(dp)
        return dp[-1]

# @lc code=end
if __name__ == "__main__":
    print(Solution().coinChange([186,419,83,408],6249))

File: golang/322.coin-change.go
/*
 * @lc app=leetcode id=322 lang=golang
 *
 * [322] Coin Change
 */
// TODO 有一个基于位运算的解决方案
// https://leetcode-cn.com/problems/coin-change/solution/ji-yu-wei-yun-suan-de-dong-tai-gui-hua-p-9m8o/
// @lc code=start
func coinChange(coins []int, amount int) int {
	dp := []int{0}
	for i := 1; i <= amount; i++ {
		min := math.MaxInt32
		for j := 0; j < len(coins); j++ {
			if i-coins[j] >= 0 && min > dp[i-coins[j]] {
				min = dp[i-coins[j]] + 1
			}
		}
		dp = append(dp, min)
	}
	if dp[len(dp)-1] == math.MaxInt32 {
		return -1
	}
	return dp[len(dp)-1]
}

// @lc code=end

